ETH-SECPM-DEC (#28)
vacp2p/rfc-index: vac/raw/decentralized-messaging-ethereum.md

### Components of the protocol

This protocol relies in 3 components: 
authenticated causal broadcast (ACB), 
decentralized group membership (DGM) and 
2-party secure messaging (2SM).

#### Authenticated causal broadcast
A causal order is a partial order relation `<` on messages. 
Two messages `m_1` and `m_2` are causally ordered, or 
`m_1` causally precedes `m_2` 
(denoted by `m_1 < m_2`), if one of the following contiditions hold:

1. `m_1` and `m_2` were sent by the same group member, and `m_1` was sent before `m_2`.
2. `m_2` was sent by a group member U, and `m_1` was received and
processed by `U` before sending `m_2`.
3. There exists `m_3` such that `m_1 < m_3` and `m_3 < m_2`.

Causal broadcast requires that before processing `m`, 
a group member must process all preceding messages `{m' | m' < m}`.

The causal broadcast module used in this protocol authenticates the sender of each message, 
as well as its causal ordering metadata, using a digital signature under the sender’s identity key. 
This prevents a passive adversary from impersonating users or affecting causally ordered delivery.

#### Decentralized group membership

> This protocol assumes the existence of a DGM function (denoted as DGM) 

Reviewer: ```suggestion
This protocol assumes the existence of a decentralized group membership function (denoted as DGM) 
```

that takes a set of membership change messages and their causal order relantionships, 
and returns the current set of group members’ IDs. 
It needs to be deterministic and depend only on causal order, and not exact order.

#### 2-party secure messaging (2SM)
This protocol makes use of bidirectional 2-party secure messaging schemes, 
which consist of 3 algorithms: `2SM-Init`, `2SM-Send` and `2SM-Receive`.

##### 2SM-Init
This function takes two IDs as inputs: 

> ID1 representing the local user and ID2 representing the other party. 

Reviewer: ```suggestion
`ID1` representing the local user and `ID2` representing the other party. 
```

> It returns an initial protocol state sigma`.

Reviewer: ```suggestion
It returns an initial protocol state `sigma`.
```

Additionally, this function generates a new key pair for the local user, 
storing the secret key in `sigma.mySks` and sending the public key. 
Similarly, it generates a new key pair for the other party, 
sending the secret key (encrypted) and storing the public key in `sigma.otherPk`.

```
sigma.mySks[sigma.nextIndex], myNewPk) = PKE-Gen()
(otherNewSk, otherNewPk) = PKE-Gen()
plaintext = (m, otherNewSk, sigma`.nextIndex, myNewPk)
msg = (PKE-Enc(sigma.otherPk, plaintext), sigma.otherPkSender, sigma.otherPkIndex)
sigma.nextIndex++
(sigma.otherPk, sigma.otherPkSender, sigma.otherPkIndex) = (otherNewPk, "me", empty_string)
return (sigma`, msg)
```

#### 2SM-Receive

This function utilizes the metadata of the message `c` to determine which secret key to utilize for decryption, 
assigning it to `sk`. 
If the secret key corresponds to one generated by ourselves, 
that secret key along with all keys with lower index are deleted. 
This deletion is indicated by `sigma.mySks[≤ keyIndex] = empty_string`. 
Subsequently, the new public and secret keys contained in the message are stored.

```

> (ciphertext, keySender, keyIndex) = c

Reviewer: Maybe we need to format it as code? I mean add tabs under if-else

if keySender = "other" then 
sk = sigma.mySks[keyIndex] 
sigma.mySks[≤ keyIndex] = empty_string
else sk = sigma.receivedSk
(m, sigma.receivedSk, sigma.otherPkIndex, sigma.otherPk) = PKE-Dec(sk, ciphertext)
sigma.otherPkSender = "other"
return (sigma, m)
```
### PKE Syntax

The required PKE that MUST be used is ElGamal with a 2048-bit modulus `p`. 

#### Parameters

The following parameters must be used:

```
p = 308920927247127345254346920820166145569
g = 2
```

#### PKE-KGen

Each user `u` MUST do the following:

```
PKE-KGen():

> a = randint(1, p-2)

Reviewer: Little suggestion. Maybe it's better to choose a in interval [2, p-2] for all further computation regarding random value it ElGamal. I mean DLP is not a problem if our secret is 1 :) 

That's true that in original white paper interval is [1;p-2]. But in later work regarding prevent attack I saw that interval [2;p-2] is better

pk = (p, g, g^a)
sk = a
return (pk, sk)
```

#### PKE-Enc

A user `v` encrypting a message `m` for `u` MUST follow these steps:

```
PKE-Enc(pk):
k = randint(1, p-2)
eta = g^k % p
delta = m * (g^a)^k % p
return ((eta, delta))
```

#### PKE-Dec

The user `u` recovers a message `m` from a ciphertext `c` by performing the following operations:

```
PKE-Dec(sk):
mu = eta^(p-1-sk) % p

> return ((mu` * delta) % p)

Reviewer: ```suggestion
return ((mu * delta) % p)
```

Or I miss sth about mu` ?

```

### DCGKA Syntax
#### Auxiliary functions

There exist 6 functions that are auxiliary for the rest of components of the protocol, namely:

> ##### init

Reviewer: ```suggestion
#### init
```

This function takes an `ID` as input and returns its associated initial state, denoted by `gamma`:

```
gamma.myId = ID
gamma.mySeq = 0
gamma.history = empty
gamma.nextSeed = empty_string
gamma.2sm[·] = empty_string
gamma.memberSecret[·, ·, ·] = empty_string
gamma.ratchet[·] = empty_string
return (gamma)
```

#### encrypt-to

Upon reception of the recipient’s `ID` and a plaintext, 
it encrypts a direct message for another group member. 
Should it be the first message for a particular `ID`, 
then the `2SM` protocol state is initialized and stored in `gamma.2sm[recipient.ID]`. 
One then uses `2SM_Send` to encrypt the message and store the updated protocol in `gamma`.

```

> if gamma.2sm[recipient_ID] = empty_string then

Reviewer: Here I would also suggest to add tab inside `if`

gamma.2sm[recipient_ID] = 2SM_Init(gamma.myID, recipient_ID)
(gamma.2sm[recipient_ID], ciphertext) = 2SM_Send(gamma.2sm[recipient_ID], plaintext)
return (\gamma, ciphertext)
```

#### decrypt-from

After receiving the sender’s `ID` and a ciphertext, 
it behaves as the reverse function of `encrypt-to` and has a similar initialization:

```
if gamma.2sm[sender_ID] = empty_string then
gamma.2sm[sender_ID] = 2SM_Init(gamma.myID, sender_ID)
(gamma.2sm[sender_ID], plaintext) = 2SM_Receive(gamma.2sm[sender_ID], ciphertext)
return (gamma, plaintext)
```

#### update-ratchet

> This function generates the next update secret `I` for the group member `ID`. 

Reviewer: In this section I suggest to add an index for `update secret` value (e.g `I_j` or `I_next` or sth else). I think It would give better understanding for those one who go through function description that it is the same value as in start of rfc where it has indexes

The ratchet state is stored in `gamma.ratchet[ID]`. 
It is required to use a HMAC-based key derivation function HKDF to combine the ratchet state with an input, 
returning an update secret and a new ratchet state.

```
(updateSecret, gamma.ratchet[ID]) = HKDF(gamma.ratchet[ID], input)
return (gamma, updateSecret)
```

#### member-view

This function calculates the set of group members based on the most recent control message sent by the specified user `ID`. 
It filters the group membership operations to include only those observed by the specified `ID`, and 
then invokes the DGM function to generate the group membership.

```

> ops = {m in gamma.history st. m was sent or acknowledged by ID}

Reviewer: What does `st.` mean?

Author: Such that :-)

return DGM(ops)
```

#### generate-seed

This functions generates a random bit string and 
sends it encrypted to each member of the group using the `2SM` mechanism. 
It returns the updated protocol state and 

> the set of direct messages to send.

Reviewer: ```suggestion
the set of direct messages (denoted as `dmsgs`) to send.
```

Because I spend time to find difference between msg and dmsg in next functions description

```
gamma.nextSeed = random.randbytes()
dmsgs = empty
for each ID in recipients:
(gamma, msg) = encrypt-to(gamma, ID, gamma.nextSeed)
dmsgs = dmsgs + (ID, msg)
return (gamma, dmsgs)
```

### Creation of a group

A group is generated in a 3 steps procedure:

1. A user calls the `create` function and broadcasts a control message of type *create*.
2. Each receiver of the message processes the message and broadcasts an *ack* control message.
3. Each member processes the *ack* message received.

#### create
This function generates a *create* control message and 
calls `generate-seed` to define the set of direct messages that need to be sent. 
Then it calls `process-create` to process the control message for this user. 
The function `process-create` returns a tuple including an updated state gamma and 

> an update secret $I$.

Reviewer: ```suggestion
an update secret `I`.
```

```
control = (“create”, gamma.mySeq, IDs)
(gamma, dmsgs) = generate-seed(gamma, IDs)

> (gamma, _, _, I, _) = process-create(gamma, gamma.myId, gamma.mySeq, IDs, empty_string)

Reviewer: what the idea behind the pass gamma parameters with gamma itself?

return (gamma, control, dmsgs, I)
```

#### process-seed
This function initially employs `member-view` to identify the users who were part of the group when the control message was dispatched. 
Then, it attempts to acquire the seed secret through the following steps:

1. If the control message was dispatched by the local user,

> it uses the most recent invocation of `generate-seed` stored the seed secret in gamma.nextSeed.

Reviewer: ```suggestion
it uses the most recent invocation of `generate-seed` stored the seed secret in `gamma.nextSeed`.
```

> 2. If the *control* message was dispatched by another user, and

Reviewer: Not sure about formatting of word control

the local user is among its recipients, 
the function utilizes `decrypt-from` to decrypt the direct message that includes the seed secret.
3. Otherwise, it returns an *ack* message without deriving an update secret.

Afterwards, `process-seed` generates separate member secrets for each group member from the seed secret by combining the seed secret and 
each user ID using HKDF. 

> The secret for the sender of the message is stored in *senderSecret*, 

Reviewer: ```suggestion
The secret for the sender of the message is stored in `senderSecret`, 
```

> while those for the other group members are stored in gamma.memberSecret. 

Reviewer: ```suggestion
while those for the other group members are stored in `gamma.memberSecret`. 
```

The sender's member secret is immediately utilized to update their KDF ratchet and 
compute their update secret `I_sender` using `update-ratchet`. 
If the local user is the sender of the control message, 
the process is completed, and the update secret is returned. 
However, if the seed secret is received from another user, 

> an *ack* control message is constructed for broadcast, 

Reviewer: ```suggestion
an `ack` control message is constructed for broadcast, 
```

including the sender ID and sequence number of the message being acknowledged.

The final step computes an update secret `I_me` for the local user invoking the `process-ack` function.

```
recipients = member-view(gamma, sender) - {sender}
if sender =  gamma.myId then seed = gamma.nextSeed; gamma.nextSeed = empty_string
else if  gamma.myId in recipients then (gamma, seed) = decrypt-from(gamma, sender, dmsg)
else
return (gamma, (ack, ++gamma.mySeq, (sender, seq)), empty_string , empty_string , empty_string)

for ID in recipients do gamma.memberSecret[sender, seq, ID] = HKDF(seed, ID)

senderSecret = HKDF(seed, sender)

> (gamma, $I_{sender}$) = update-ratchet(gamma, sender, senderSecret)

Reviewer: ```suggestion
(gamma, I_sender) = update-ratchet(gamma, sender, senderSecret)
```

if sender = gamma.myId then return (gamma, empty_string , empty_string , I_sender, empty_string)
control = (ack, ++gamma.mySeq, (sender, seq))
members = member-view(gamma, gamma.myId)
forward = empty

> for ID in {members - (recipients + {sender})}`

Reviewer: ```suggestion
for ID in {members - (recipients + {sender})}
```

	s = gamma.memberSecret[sender, seq, gamma.myId]
	(gamma, msg) = encrypt-to(gamma, ID, s)
	forward = forward + {(ID, msg)}

(gamma, _, _, I_me, _) = process-ack(gamma, gamma.myId, gamma.mySeq, (sender, seq), empty_string)
return (gamma, control, forward, I_sender, I_me)
```

#### process-create

> This function is called by the sender and each of the receivers of the *create* control message. 

Reviewer: ```suggestion
This function is called by the sender and each of the receivers of the `create` control message. 
```

> First, it records the information from the create message in the `gamma.history+`, 

Reviewer: ```suggestion
First, it records the information from the create message in the `gamma.history+ {op}`, 
```

which is used to track group membership changes. Then, it proceeds to call `process-seed`.

```
op = (”create”, sender, seq, IDs)
gamma$.history = gamma.history + {op}
return (process-seed(gamma, sender, seq, dmsg))
```

#### process-ack
This function is called by those group members once they receive an ack message. 
In `process-ack`, `ackID` and `ackSeq` are the sender and 
sequence number of the acknowledged message. 
Firstly, if the acknowledged message is a group membership operation, 
it records the acknowledgement in `gamma.history`. 

Following this, the function retrieves the relevant member secret from `gamma.memberSecret`,
which was previously obtained from the seed secret contained in the acknowledged message.

> Finally, it updates the ratchet for the sender of the *ack* and 

Reviewer: ```suggestion
Finally, it updates the ratchet for the sender of the `ack` and 
```

returns the resulting update secret.

```
if (ackID, ackSeq) was a create / add / remove then
op = ("ack", sender, seq, ackID, ackSeq)

> gamma$.history = gamma.history + {op}`

Reviewer: ```suggestion
gamma.history = gamma.history + {op}`
```

s = gamma.memberSecret[ackID, ackSeq, sender]

> gamma$.memberSecret[ackID, ackSeq, sender] = empty_string

Reviewer: ```suggestion
gamma.memberSecret[ackID, ackSeq, sender] = empty_string
```

if (s = empty_string) & (dmsg = empty_string) then return (gamma, empty_string, empty_string, empty_string, empty_string)
if (s = empty_string) then (gamma, s) = decrypt-from(gamma, sender, dmsg)
(gamma, I) = update-ratchet(gamma, sender, s)
return (gamma, empty_string, empty_string, I, empty_string)
```

The HKDF function MUST follow RFC 5869 using the hash function SHA256.

### Post-compromise security updates and group member removal

The functions `update` and `remove` share similarities with `create`: 
they both call the function `generate-seed` to encrypt a new seed secret for each group member. 
The distinction lies in the determination of the group members using `member-view`. 
In the case of `remove`, the user being removed is excluded from the recipients of the seed secret. 

> Additionally, the control message they construct is designated with type *update* or *remove* respectively.

Reviewer: ```suggestion
Additionally, the control message they construct is designated with type `update` or `remove` respectively.
```

Likewise, `process-update` and `process-remove` are akin to `process-create`. 

> The function `process-update` skips the update of gamma.history, 

Reviewer: ```suggestion
The function `process-update` skips the update of `gamma.history`, 
```

whereas `process-remove` includes a removal operation in the history.

#### update
```
control = ("update", ++gamma.mySeq, empty_string)
recipients = member-view(gamma, gamma.myId) - {gamma.myId}
(gamma, dmsgs) = generate-seed(gamma, recipients)
(gamma, _, _, I , _) = process-update(gamma, gamma.myId, gamma.mySeq, empty_string, empty_string)
return (gamma, control, dmsgs, I)
```

#### remove
```
control = ("remove", ++gamma.mySeq, empty)

> recipients = member-view(gamma, gamma.myId) - {ID, gamma.myId}

Reviewer: Does this mean that anyone can remove a user from a group? Will there be any restrictions on this?

(gamma, dmsgs) = generate-seed(gamma, recipients)
(gamma, _, _, I , _) = process-update(gamma, gamma.myId, gamma.mySeq, ID, empty_string)
return (gamma, control, dmsgs, I)
```

#### process-update

> `return process-seed(gamma, sender, seq, dmsg)`

Reviewer: May be better do it in one style? 

```
return process-seed(gamma, sender, seq, dmsg)
```

#### process-remove
```
op = ("remove", sender, seq, removed)
gamma.history = gamma.history + {op}
return process-seed(gamma, sender, seq, dmsg)
```

### Group member addition

#### add
When adding a new group member, 
an existing member initiates the process by invoking the `add` function and 
providing the ID of the user to be added.

> This function prepares a control message marked as *add* for broadcast to the group. 

Reviewer: ```suggestion
This function prepares a control message marked as `add` for broadcast to the group. 
```

Simultaneously, it creates a welcome message intended for the new member as a direct message. 

> This *welcome* message includes the current state of the sender's KDF ratchet, 

Reviewer: ```suggestion
This `welcome` message includes the current state of the sender's KDF ratchet, 
```

encrypted using `2SM`, along with the history of group membership operations conducted so far.

```
control = ("add", ++gamma.mySeq, ID)
(gamma, c) = encrypt-to(gamma, ID, gamma.ratchet[gamma.myId])
op = ("add", gamma.myId, gamma.mySeq, ID)
welcome = (gamma.history + {op}, c)

> (gamma, _, _, $I$, _) = process-add(gamma, gamma.myId, gamma.mySeq, ID, empty_string)

Reviewer: ```suggestion
(gamma, _, _, I, _) = process-add(gamma, gamma.myId, gamma.mySeq, ID, empty_string)
```

> return (gamma, control, {(ID, welcome)}, $I$)

Reviewer: ```suggestion
return (gamma, control, (ID, welcome), I)
```

```

#### process-add
This function is invoked by both the sender and 

> each recipient of an *add* message, which includes the new group member. 

Reviewer: ```suggestion
each recipient of an `add` message, which includes the new group member. 
```

If the local user is the newly added member, 
the function proceeds to call `process-welcome` and then exits. 
Otherwise, it extends `gamma.history` with the `add` operation.

> Line 5 determines whether the local user was already a group member at the time the "add" message was sent; 

Reviewer: Here and below in this section is a line by line description and perhaps then we should add them to the pseudo code?

> Line 5 determines whether the local user was already a group member at the time the "add" message was sent; 

Reviewer: ```suggestion
Line 5 determines whether the local user was already a group member at the time the `add` message was sent; 
```

this condition is typically true but may be false if multiple users were added concurrently.

On lines 6 to 8, the ratchet for the sender of the *add* message is updated twice. 
In both calls to `update-ratchet`, 
a constant string is used as the ratchet input instead of a random seed secret.

The value returned by the first ratchet update is stored in `gamma.memberSecret` as the added user’s initial member secret. 

> The result of the second ratchet update becomes I_sender, 

Reviewer: ```suggestion
The result of the second ratchet update becomes `I_sender`, 
```

> the update secret for the sender of the *add* message. 

Reviewer: ```suggestion
the update secret for the sender of the `add` message. 
```

On line 10, if the local user is the sender, the update secret is returned.

> If the local user is not the sender, an acknowledgment for the *add* message is required. 

Reviewer: ```suggestion
If the local user is not the sender, an acknowledgment for the `add` message is required. 
```

> Therefore, on line 11, a control message of type *add-ack* is constructed for broadcast. 

Reviewer: ```suggestion
Therefore, on line 11, a control message of type `add-ack` is constructed for broadcast. 
```

Subsequently, in line 12 the current ratchet state is encrypted using `2SM` to generate a direct message intended for the added user, 
allowing them to decrypt subsequent messages sent by the sender. 

> Finally, in lines 13 to 15, `process-add-ack` is called to calculate the local user’s update secret (I_me), 
> which is then returned along with I_sender.

Reviewer: ```suggestion
Finally, in lines 13 to 15, `process-add-ack` is called to calculate the local user’s update secret (`I_me`), 
which is then returned along with `I_sender`.
```

```
if added = gamma.myId then return process-welcome(gamma, sender, seq, dmsg)
op = ("add", sender, seq, added)
gamma.history = gamma.history + {op}
if gamma.myId in member-view(gamma, sender) then

> 	(gamma, s) = update-ratchet(\gamma, sender, "welcome")

Reviewer: ```suggestion
	(gamma, s) = update-ratchet(gamma, sender, "welcome")
```

	gamma.memberSecret[sender, seq, added] = s
	(gamma, I_sender) = update-ratchet(gamma, sender, "add")
else I_sender = empty_string
if sender = gamma.myId then return (gamma, empty_string, empty_string, I_sender, empty_string)
control = ("add-ack", ++gamma.mySeq, (sender, seq))
(gamma, c) = encrypt-to(gamma, added, ratchet[gamma.myId])
(gamma, _, _, I_me, _) = process-add-ack(gamma, gamma.myId, gamma.mySeq, (sender, seq), empty_string)

> return (gamma, control, {(added, c)}, I_sender, I_me)

Reviewer: ```suggestion
return (gamma, control, (added, c), I_sender, I_me)
```

```

#### process-add-ack

> This function is invoked by both the sender and each recipient of an *add-ack* message, 
> including the new group member. 
> Upon lines 1–2, the acknowledgment is added to `gamma.history`, 
> mirroring the process in `process-ack`. 
> If the current user is the new group member, 
> the *add-ack* message includes the direct message constructed in `process-add`; 
> this direct message contains the encrypted ratchet state of the sender of the *add-ack*, 
> then it is decrypted on lines 3–5.
> 
> Upon line 6, a check is performed to check if the local user was already a group member at the time the *add-ack* was sent. 
> If affirmative, a new update secret `I` for the sender of the *add-ack* is computed on line 7 by invoking `update-ratchet` with the constant string *add*.
> 
> In the scenario involving the new member, 
> the ratchet state was recently initialized on line 5. T
> This ratchet update facilitates all group members, including the new addition, 
> to derive each member’s update by obtaining any update secret from before their inclusion.

Reviewer: ```suggestion
This function is invoked by both the sender and each recipient of an `add-ack` message, 
including the new group member. 
Upon lines 1–2, the acknowledgment is added to `gamma.history`, 
mirroring the process in `process-ack`. 
If the current user is the new group member, 
the `add-ack` message includes the direct message constructed in `process-add`; 
this direct message contains the encrypted ratchet state of the sender of the `add-ack`, 
then it is decrypted on lines 3–5.

Upon line 6, a check is performed to check if the local user was already a group member at the time the `add-ack` was sent. 
If affirmative, a new update secret `I` for the sender of the `add-ack` is computed on line 7 by invoking `update-ratchet` with the constant string `add`.

In the scenario involving the new member, 
the ratchet state was recently initialized on line 5.
This ratchet update facilitates all group members, including the new addition, 
to derive each member’s update by obtaining any update secret from before their inclusion.
```

```
op = ("ack", sender, seq, ackID, ackSeq)
gamma$.history = gamma.history + {op}
if dmsg != empty_string then
	(gamma, s) = decrypt-from(gamma, sender, dmsg)
	gamma.ratchet[sender] = s
if gamma.myId in member-view(gamma, sender) then

> 	(gamma, $I$) = update-ratchet(gamma, sender, "add")

Reviewer: ```suggestion
	(gamma, I) = update-ratchet(gamma, sender, "add")
```

	return (gamma, empty_string, empty_string, I, empty_string)
else return (gamma, empty_string, empty_string, empty_string, empty_string)
```

#### process-welcome
This function serves as the second step called by a newly added group member. 

> In this context, *adderHistory* represents the adding user’s copy of `gamma.history` sent in their welcome message, 

Reviewer: ```suggestion
In this context, `adderHistory` represents the adding user’s copy of `gamma.history` sent in their welcome message, 
```

which is utilized to initialize the added user’s history. 
Here, `c` denotes the ciphertext of the adding user’s ratchet state, 
which is decrypted on line 2 using `decrypt-from`.

Once `gamma.ratchet[sender]` is initialized, 

> `update-ratchet` is invoked twice on lines 3 to 5 with the constant strings *welcome* and *add* respectively. 

Reviewer: ```suggestion
`update-ratchet` is invoked twice on lines 3 to 5 with the constant strings `welcome` and `add` respectively. 
```

These operations mirror the ratchet operations performed by every other group member in `process-add`. 
The outcome of the first `update-ratchet` call becomes the first member secret for the added user, 

> while the second call returns I_sender, the update secret for the sender of the add operation.

Reviewer: ```suggestion
while the second call returns `I_sender`, the update secret for the sender of the add operation.
```

Subsequently, the new group member constructs an *ack* control message to broadcast on line 6 and 
calls `process-ack` to compute their initial update secret I_me. 
The function `process-ack` reads from `gamma.memberSecret` and 
passes it to `update-ratchet`. 
The previous ratchet state for the new member is the empty string `empty`, as established by `init`, 
thereby initializing the new member’s ratchet. 

> Upon receiving the new member’s *ack*, 

Reviewer: ```suggestion
Upon receiving the new member’s `ack`, 
```

every other group member initializes their copy of the new member’s ratchet in a similar manner.

By the conclusion of `process-welcome`, 
the new group member has acquired update secrets for themselves and the user who added them. 
The ratchets for other group members are initialized by `process-add-ack`.

```
gamma.history = adderHistory
(gamma, gamma.ratchet[sender]) = decrypt-from(gamma, sender, c)
(gamma, s) = update-ratchet(gamma, sender, "welcome")

> gamma$.memberSecret[sender, seq, gamma.myId] = s

Reviewer: ```suggestion
gamma.memberSecret[sender, seq, gamma.myId] = s
```

4. The user authenticates their identity by signing in with their wallet.
5. Upon successful authentication, the user's identity is confirmed or approved.
6. The website grants access to data specific to the authenticated user.

#### Our approach
The idea in the [DCGKA](https://eprint.iacr.org/2020/1281) setting closely resembles the procedure outlined in SIWE. Here: 

1. The server corresponds to user D1,
who initiates a request (instead of generating a nonce) to obtain the public key of user D2. 
2. Upon receiving the request, the wallet of D2 send the request to the user, 
3. User D2 receives the request from the wallet, and decides whether accepts or rejects.
4. The wallet and responds with a message containing the requested public key in case of acceptance by D2. 

This message may be signed, allowing D1 to verify that the owner of the received public key is indeed D2.

### Multi-device setting
One may see the set of devices as a group and create a group key for internal communications. 
One may use treeKEM for instance, 
since it provides interesting properties like forward secrecy and post-compromise security. 
All devices share the same `ID`, 

> which is hold by one of them, and from other user’s point of view, they would look as a single user.

Reviewer 2: ```suggestion
which is held by one of them, and from other user’s point of view, they would look as a single user.
```

Using servers, like in the paper [Multi-Device for Signal](https://eprint.iacr.org/2019/1363), should be avoided; 
but this would imply using a particular device as receiver and broadcaster within the group. 
There is an obvious drawback which is having a single device working as a “server”. 
Should this device be attacked or without connection, there should be a mechanism for its revocation and replacement.

Another approach for communications between devices could be using the keypair of each device. 
This could open the door to use UPKE, since keypairs should be regenerated frequently.

Each time a device sends a message, either an internal message or an external message, 
it needs to replicate and broadcast it to all devices in the group.

> ![Captura de pantalla 2024-03-21 a les 12.07.23.png](https://prod-files-secure.s3.us-west-2.amazonaws.com/1518abd9-c08f-4989-93c1-96525e62bce5/9144c2dd-aa6c-4f7c-9acc-f9dc7f18661d/Captura_de_pantalla_2024-03-21_a_les_12.07.23.png)

Reviewer: This picture is also unavailable

To prevent other devices from sending messages outside their group, a requirement should be signing each message. 
The keys associated to the `ID` should only be in control of the leader device.

The leader device is in charge of setting the keys involved in the DCGKA. 
This information must be replicated within the group to make sure it is updated.

To detect missing messages or potential misbehavior, messages must include a counter.

### Using UPKE

Managing the group of devices of a user can be done either using a group key protocol such as treeKEM or 
using the keypair of each device. 
Setting a common key for a group of devices under the control of the same actor might be excessive, 
furthermore it may imply some of the problems one can find in the usual setting of a group of different users; 
for example: one of the devices may not participate in the required updating processes, representing a threat for the group.

The other approach to managing the group of devices is using each device’s keypair, 
but it would require each device updating these materia frequently, something that may not happens. 

> [Updatable public-key encryption](https://eprint.iacr.org/2022/068) is a form of asymetric cryptography 

Reviewer 2: ```suggestion
[UPKE](https://eprint.iacr.org/2022/068) is a form of asymetric cryptography 
```

Reviewer 2: Or we can change the reference [UPKE](https://eprint.iacr.org/2022/068) as [Updatable public-key encryption](https://eprint.iacr.org/2022/068)

where any user can update any other user’s key pair by running an update algorithm with (high-entropy) private coins. 
Any sender can initiate a *key update* by sending a special update ciphertext. 
This ciphertext updates the receiver’s public key and also, once processed by the receiver, will update their secret key.

To the best of my knowledge, 

> there exists several efficient constructions both [classic](https://eprint.iacr.org/2019/1189) (based in the DH assumption) and 

Reviewer 2: ```suggestion
there exists several efficient constructions both [UPKE from ElGamal](https://eprint.iacr.org/2019/1189) (based in the DH assumption) and 
```

> [postquantum]((https://eprint.iacr.org/2023/1400)) (based in lattices). 

Reviewer 2: ```suggestion
[UPKE from Lattices]((https://eprint.iacr.org/2023/1400)) (based in lattices). 
```
